shell/platform/android/io/flutter/embedding/engine/systemchannels/SettingsChannel.java

```diff
@@ -17,13 +21,27 @@ public class SettingsChannel {
   private static final String BRIEFLY_SHOW_PASSWORD = "brieflyShowPassword";
   private static final String ALWAYS_USE_24_HOUR_FORMAT = "alwaysUse24HourFormat";
   private static final String PLATFORM_BRIGHTNESS = "platformBrightness";
+  private static final String CONFIGURATION_GENERATION = "configurationId";
 
   @NonNull public final BasicMessageChannel<Object> channel;
 
   public SettingsChannel(@NonNull DartExecutor dartExecutor) {
     this.channel = new BasicMessageChannel<>(dartExecutor, CHANNEL_NAME, JSONMessageCodec.INSTANCE);
   }
 
+  public static DisplayMetrics getPastDisplayMetrics(int configGeneration) {
+    final LinkedList<SentConfiguration> pastConfigurations = SentConfiguration.sentConfigurations;
+    while (pastConfigurations.size() > 0
+        && pastConfigurations.getFirst().generationNumber != configGeneration) {
+      pastConfigurations.remove();
```

Comment:
My understanding is that `SentConfiguration.sentConfigurations` is modified from two separate threads: the code here is executed on the UI thread, and the code around line 129 is executed on the platform thread. Are these operations thread save? Looking at it naively, I would say they are not and if thread switching happens at a bad time we could end up with data loss?

Scenario: Let's say SentConfiguration.sentConfigurations looks like this: A -> B and the platform thread is still waiting for acknowledgement of B, which the UI thread has already received, but the ack has not been processed on the platform thread yet. For simplicity, let's assume that we only have one core to execute threads:

The UI thread is running and now asks for the display metrics of generation B. `getPastDisplayMetrics` above will now think it can delete A from the list. However, right before line 36 executes, the threads switch and the platform thread executes. It now processes the ack and proceeds to delete generation A. Some time after that, the threads switch again, the UI thread runs and it can finally execute line 36, which will now incorrectly delete generation B. 🔥 🔥 🔥 

Ideally, we would find a solution to not require complicated thread synchronization mechanisms... Would it be possible to somehow send the data required to do the non-linear scaling over to the UI thread with the rest of the platform metrics so we don't have to call back and deal with this race condition at all?

Reply:
(Or if that's not an option and we need to do this complicated dance with generations we could probably just change it to have the UI thread only read this data, but never modify it)

Reply:
The reply callback is called on the platform thread I think: https://github.com/flutter/engine/blob/9babd40c04b16659d44fa2caa0d1c2cc653c47ca/shell/platform/android/platform_message_response_android.cc#L27

Reply:
There's going to be a new AndroidX API that takes the unscaled size and the scale factor, but I don't think it's available yet and I think it adds some additional AndroidX dependencies.

Reply:
I'll update `getPastDisplayMetrics` so it doesn't modify the linked list and add a RW lock.